### PR TITLE
 Add a crate for shared rusqlite utilities, and use it in `logins-sql` and `places`

### PR DESCRIPTION
**Pull request by thomcc**

### CODE REVIEW

**Comment by thomcc** [comment]
The biggest open question I have here is whether or not we can do something around schema management/upgrade. I'm also going to rename it and mess with the directory structure.

**Comment by thomcc** [comment]
I'd also like to factor out shared code for FFI stuff, but will leave that for after (and it's unrelated to SQL). Currently there's some duplication between Fxa and logins (in particular, around errors), and I started a little on the Places FFI, which will have the same issue as well.

File: components/support/sql/src/conn_ext.rs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    /// Get a cached or uncached statement based on a flag.
+    fn prepare_maybe_cached<'conn>(&'conn self, sql: &str, cache: bool) -> SqlResult<MaybeCached<'conn>> {
+        MaybeCached::prepare(self.conn(), sql, cache)
+    }
+
+    /// Execute all the provided statements.
+    fn execute_all(&self, stmts: &[&str]) -> SqlResult<()> {
+        let conn = self.conn();
+        for sql in stmts {
+            conn.execute(sql, &[])?;
```
Comment (linacambridge) [inline components/support/sql/src/conn_ext.rs]: Would it make sense to use `conn.execute_batch` here?

File: components/support/sql/src/conn_ext.rs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        E: From<rusqlite::Error>,
+        F: FnOnce(&Row) -> Result<T, E>,
+    {
+        Ok(self.try_query_row(sql, params, mapper, cache)?
+            .ok_or(rusqlite::Error::QueryReturnedNoRows)?)
+    }
+
+    // This should probably have a longer name...
+    /// Like `query_row_and_then_named` but returns None instead of erroring if no such row exists.
+    fn try_query_row<T, E, F>(
```
Comment (linacambridge) [inline components/support/sql/src/conn_ext.rs]: I think this is a fine name! Maybe `try_query_row_named`, if you want, but it's pretty clear what it does.

File: components/support/sql/src/each_chunk.rs [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+where
+    T: ToSql + 'a,
+    F: FnMut(&[&dyn ToSql], usize) -> Result<(), E>,
+{
+    each_sized_chunk_mapped(items, default_max_variable_number(), |t| t as &dyn ToSql, do_chunk)
+}
+
+/// A version of `each_chunk` for the case when the conversion to `to_sql` requires an custom
+/// intermediate step. For example, you might want to grab a property off of an arrray of records
+pub fn each_chunk_mapped<'a, T, E, Mapper, DoChunk>(
```
Comment (linacambridge) [inline components/support/sql/src/each_chunk.rs]: We could maybe implement this on `ConnExt`, which would also let us query that specific connection's limit...but I think this makes sense, too. I also agree that we don't need to support lowering the limit at runtime. 👍

**Pull request merged**